### PR TITLE
factory: use customised deep compare

### PR DESCRIPTION
**Pull request by bergwolf**
reflect.DeepEqual would return false when comparing nil map/slice with
empty map/slice. We would want to return success in such case, since it
is possible for upper layers to send these kind of configs.

Fixes: #844

### CODE REVIEW

File: virtcontainers/factory/factory.go [hunk 1]
```diff
@@ -100,6 +100,71 @@ func resetHypervisorConfig(config *vc.VMConfig) {
 	config.ProxyConfig = vc.ProxyConfig{}
 }
 
+func compareStruct(foo, bar reflect.Value) bool {
+	for i := 0; i < foo.NumField(); i++ {
```
Comment (jodh-intel) [inline virtcontainers/factory/factory.go]: Missing check on field counts for `foo` and `bar`? They might compare as the same but if `bar` has extra fields that would be incorrect surely?

Comment (bergwolf) [reply virtcontainers/factory/factory.go]: They have to pass `foo.Type() == bar.Type()` so they must be of the same struct type.

Comment (jodh-intel) [reply virtcontainers/factory/factory.go]: Ah - good point!

**Comment by jodh-intel** [comment]
lgtm

**Comment by WeiZhang555** [comment]
LGTM

CI retriggered, let's see if there's any real issue.

**Comment by jodh-intel** [comment]
Still seeing network timeouts.

/test

**Comment by bergwolf** [comment]
CI green now.

**Pull request merged**